### PR TITLE
[Embedding] Support to avoid preloaded IDs being eliminated in multi-tier embedding's cache.

### PR DESCRIPTION
**Pull request by lixy9474**

### CODE REVIEW

File: tensorflow/core/framework/embedding/cache.h [hunk 1]
```diff
@@ -51,6 +64,56 @@ class BatchCache {
   int64 num_miss;
 };
 
+template<class K>
+class PrefetchNode {
+ public:
+  PrefetchNode(): key_(-1), ref_count_(1) {}
+  PrefetchNode(K id): key_(id), ref_count_(1) {}
```
Comment (candyzone) [inline tensorflow/core/framework/embedding/cache.h:71]: explicit

File: tensorflow/core/framework/embedding/cache.h [hunk 2]
```diff
@@ -99,8 +162,10 @@ class LRUCache : public BatchCache<K> {
     return i;
   }
 
-  void add_to_rank(const K* batch_ids, size_t batch_size) {
-    mutex_lock l(mu_);
+  void add_to_rank(const K* batch_ids, size_t batch_size,
+                   bool is_locked=false) {
```
Comment (candyzone) [inline tensorflow/core/framework/embedding/cache.h:166]: rename to use_locking

File: tensorflow/core/framework/embedding/cache.h [hunk 3]
```diff
@@ -99,8 +162,10 @@ class LRUCache : public BatchCache<K> {
...
+    if (!is_locked)
```
Comment (candyzone) [inline tensorflow/core/framework/embedding/cache.h:167]: 花括号

File: tensorflow/core/framework/embedding/cache.h [hunk 4]
```diff
@@ -123,14 +188,60 @@ class LRUCache : public BatchCache<K> {
         BatchCache<K>::num_miss++;
       }
     }
+    if (!is_locked)
+      mu_.unlock();
```
Comment (candyzone) [inline tensorflow/core/framework/embedding/cache.h:192]: 避免这种调用，代码分支多了不好维护，用对象的机制

**Review by liutongxuan — APPROVED** [review]

**Pull request merged**